src/dj.erl

```diff
@@ -1,222 +1,810 @@
+%%
+%% @see float/0
+%% @see pos_integer/0
+%% @see integer/0
+%% @see non_neg_integer/0
+-spec neg_integer() -> decoder(neg_integer()).
+neg_integer() ->
+  fun (Json) when is_integer(Json), Json < 0 ->
+      {ok, Json};
+      (Json) ->
+      unexpected_type_error(neg_integer, Json)
+  end.
+
+%% @doc Decodes a positive JSON integer as a `non_neg_integer()'
+%%
+%% @see float/0
+%% @see pos_integer/0
+%% @see neg_integer/0
+%% @see integer/0
+-spec non_neg_integer() -> decoder(non_neg_integer()).
+non_neg_integer() ->
+  fun (Json) when is_integer(Json), Json >= 0 ->
+      {ok, Json};
+      (Json) ->
+      unexpected_type_error(non_neg_integer, Json)
+  end.
+
+%% @doc Decodes a JSON number as a `float()'
+%%
+%% Note that JSON does not have a separate floating point type. As such,
+%% integers in JSON will be cast to floats by this function.
+%%
+%% ```
+%%    {ok, 123.0} = dj:decode(<<"123">>, dj:float()).
+%% '''
+%%
+%% @see integer/0
+%% @see pos_integer/0
+%% @see neg_integer/0
+%% @see non_neg_integer/0
+-spec float() -> decoder(float()).
+float() ->
+  fun (Json) when is_float(Json)   -> {ok, Json};
+      (Json) when is_integer(Json) -> {ok, float(Json)};
+      (Json)                       -> unexpected_type_error(float, Json)
+  end.
+
+%% @equiv null(null)
+-spec null() -> decoder(null).
+null() ->
+  null(null).
+
+%% @doc Decodes `null' into an arbitrary value.
+%%
+%% This can be used to convert `null' to a specific value, like `undefined' or a
+%% default value that makes sense for your application.
+%%
+%% ```
+%%    {ok, foo} = dj:decode(<<"null">>, dj:null(foo)).
+%% '''
+%%
+%% @see null/0
+-spec null(V) -> decoder(V).
+null(V) ->
+  fun (null) -> {ok, V};
+      (Json) -> unexpected_type_error(null, Json)
+  end.
+
+%% @doc Decodes a JSON `true' or `false' to a `boolean()'
+%%
+%% ```
+%%    {ok, true} = dj:decode(<<"true">>, dj:boolean()).
+%%    {ok, false} = dj:decode(<<"false">>, dj:boolean()).
+%%    {error, _} = dj:decode(<<"null">>, dj:boolean()).
+%% '''
+-spec boolean() -> decoder(boolean()).
+boolean() ->
+  fun (Json) when is_boolean(Json) -> {ok, Json};
+      (Json)                       -> unexpected_type_error(boolean, Json)
+  end.
+
+%% @doc Extracts a raw `jsx:json_term()'
+-spec value() -> decoder(jsx:json_term()).
+value() ->
+  fun (X) -> {ok, X} end.
+
+%% @doc Decodes a JSON value to an `atom()'
+%%
+%% If the JSON value is `true', `false' or `null', this returns an erlang atom
+%% `true', `false' or `null'. If the JSON value is a string,
+%% `binary_to_atom(Json, utf8)' is used to turn it into an atom.
+%%
+%% <strong>NOTE</strong>: Be careful with this function, as it may be used to
+%% force erlang to create many, many new atoms, to the point of running out of
+%% memory.
+%%
+%% @see existing_atom/0
+%% @see atom/1
+-spec atom() -> decoder(atom()).
+atom() ->
+  chain( value()
+       , fun (Json) when is_atom(Json)   -> succeed(Json);
```

Comment:
I sort of wanted to implement the convenience decoders with the public API of the module, built on top of the lower level functionality.

This is also reflected in the error messages, which end up as `{custom, {not_an_atom, V}}` rather than `{unexpected_type, bool_null_or_binary, V}`.

Perhaps I should more clearly section this off with a comment, to make it clear that the convenience decoders are only there for convenience, and don't require anything you couldn't do if `decoder` was an opaque type.

Reply:
Ah, that makes sense